### DISCUSSION TITLE
ntptime (peterhinch)

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#13694] [Libraries & Drivers] [answered]
After running this version without any problems for >6months, I am now experience a low-success rate at getting NTPtime.  Both ESP32 and ESP32 SPIRAM variants.

I have tried extending the retry time out to 70 seconds to no apparent improvement.

Just wondering if other people are having problems getting NTPtime recently.

**Comment by beetlegigg** [comment]
I have a rpi picoW board that runs the Peter Hinch NTPtime every night so that the time never drifts on its internal RTC and it appears to be still running and showing the correct time on my display screen.  Checking my code I see the function that gets the NTP time can return True or False, but the calling of that function has been put in a loop to await a True response so I cant be sure that it did not loop a few times before getting the time but its not stuck in an endless loop.  So its appears OK for a rpi pico at least.  However the MP version on that board has been left at V1.20

**Reply by davefes** [reply]
Because getting NTP time has been my main source of system failure I have modified settime() as follows:

```py
def settime():
    import machine
    import time

    t = time()

  # if you don't get NTP time just log it and move on
  # the next day you get another chance
    if t <= 0:
        try:
            with open('errors.txt', 'a') as outfile:
                outfile.write('something went wrong getting UTC' + '\n')
        except OSError as err:
            pass

#        time.sleep(70)  # the min polling period for pool.ntp.org is 64 seconds
#        machine.reset()

    tm = gmtime(DS.localtime(t))
    machine.RTC().datetime((tm[0], tm[1], tm[2], tm[6] + 1, tm[3], tm[4], tm[5], 0))
```
For the last few weeks I have been getting multiple fails ('something went wrong getting UTC') so much so that my `errrors.txt` filled-up and locked-up the ESP32 SPIRAM.  So, now I try once and have another go the next day.  Fortunately, I am not putting these units to light or deepsleep so they will keep reasonable time if several NTP updates are missed.

I don't want to run a local time server, maybe I should put a DS3231 in each device and sync it once a month/year!

**Reply by zcattacz** [reply]
Do you connect wifi on demand or maintains a live link ? If there is a network issue, resolve the network issue instead of turning it into a NTP issue ? How about logging pings to the gateway or dns test first before initiating the ntp request?

**Reply by glenn20** [reply]
@davefs - if you are bringing your network interface up to run `settime()`{as suggested by @zcattacz), could it be that you are running into this issue reported by @peterhinch : https://github.com/orgs/micropython/discussions/13688. ie, your interface may not be fully initialised when you call ntptime.

**Reply by davefes** [reply]
@zcattacz I connect on demand.  Reverting to the official ntptime on a generic ESP32 works.  I will consider pinging 8.8.8.8  before continuing.
@glenn20 I am calling ntptime immediately after connecting, so thanks for the heads-up.  Any other suggestions, other then pinging the nameserver?  Maybe, just needs a few seconds delay.

**Reply by zcattacz** [reply, marked as answer]
I adapted uping for async single packet ping https://github.com/zcattacz/uping_as . You can have a try if needed. It won't lock when waiting for response. Domain resolving is also kind of ping to your DNS server, it might lock, I am not sure.

**Reply by davefes** [reply]
@zcattacz thank you for that, will investigate tomorrow.  I have been running a test for the last hour with Peter Hinch's version every 70 seconds with one error on generic v1.22.1

**Reply by peterhinch** [reply]
Calling an NTP server too frequently can lead to it rejecting attempts.

**Reply by davefes** [reply]
I had been using 20 second retries until reading another comment that 64 seconds was the minimum, so made it 70 seconds.  I could change to one try a day and accept some small drift as these ESP32s don't go to sleep.

**Reply by davefes** [reply]
Thought I would start with uPing.  I get the odd missing packet on `8.8.8.8` and also `pool.ntp.org` and sometimes 0 packets.  So, how can I be sure if I query `pool.ntp.org` at any particular instance that he will respond?

I currently try accessing `pool.ntp.org` 5 times with 70 second wait periods, then `machine.reset()`.  That even fails sometimes.

Maybe, a DS3231, accept +/- 1 minute/year and do a annual "time-sync".

**Comment by beetlegigg** [comment]
Just a thought.  Maybe your router has an inbuilt time server.  My Fritzbox router does and once I realised it (for some reason I was reading the manual) I usually used this instead of reaching out to an ntp server on the web.

**Reply by davefes** [reply]
Thanks for the prompt, however I am pretty sure that my HUAWEI-E8231 dongle does not have this feature.

It would be interesting to know what strategy the FritzBox router uses to get it's NTP time ... then maybe I could use that on the ESP32!

**Reply by zcattacz** [reply]
I am curious about your latency, do you get similar latency on your computer ? 
One of the issue that almost got me to give up is the coro task's cooperative nature. I have tasks with dead loops racing too hard under certain condition that exhausted all the resource my other tasks need. And mpy sometimes rests sometimes deadlock after days of running. Mine is only one of those extreme cases, I guess.

As to ntp maybe the ntp lib can introduce some resilient (but less accurate) mode by shooting several request packets in a row and take the first valid response.

**Reply by davefes** [reply]
About 400-500ms on the ESP32 and 30-70ms on the laptop

**Reply by davefes** [reply]
The laptop is on a cellular modem.  The ESP32 is on the HUWAEI dongle.

**Reply by davefes** [reply]
Note there are restrictions on how fast you can poll a NTP server. I often find that if it fails once then it will fail 5 times until you go away and do something else.

**Reply by davefes** [reply]
Now it is 10ms on the ESP32.

**Reply by zcattacz** [reply]
Does the ~10ms latency stick when it happens? If you can repeat the HUWAEI dongle madness on pc, maybe try reinitializing the dongle link until you have a good one. 
Place it in a different direction, orientation, location? Keep the antenna away from metal structure? Look for the magic massage :-p

I understand it will block. Shots in a row, doesn't change the fact that next try needs to be paced, I just hope it might increase the chance our packet successfully reach the server, since your link is losing packets.

**Reply by davefes** [reply]
> Does the ~10ms latency stick when it happens?

Not sure what you mean.  There doesn't seen to be any relationship between the fails and the latency.

However ... I have two of these dongles mounted on a fibreglass pole on the roof fed by long USB cables ... so I tried pinging 8.8.8.8 with the laptop on both networks.  One dongle no problem.  The other dongle (the one I am having the problem with) even after 15-30 seconds gives me:

```
64 bytes from 8.8.8.8: icmp_seq=164 ttl=247 time=73.7 ms
From 192.168.10.1 icmp_seq=165 Destination Net Unreachable
From 192.168.10.1 icmp_seq=167 Destination Net Unreachable
From 192.168.10.1 icmp_seq=168 Destination Net Unreachable
64 bytes from 8.8.8.8: icmp_seq=169 ttl=247 time=151 ms
```
This dongle is slightly lower and "pointing' in a different direction than the one that seems to work better.  You seem to have identified a problem I have been struggling with for several years now.

I also have another cellular modem but at 16Watts/hour I don't leave it running 24/7 (off-grid).

I think I now know what I am up against ... thank you.

**Comment by peterhinch** [comment]
> Maybe your router has an inbuilt time server.

There is a lot to be said for doing this. I have a Raspberry Pi running 24/7. Among other tasks this runs one.
